File: 第14课 课件+四种语言模板+代码/【并查集模板】lc547 省份数量.py
class DisjointSet {
public:
    DisjointSet(int n) {
        fa = vector<int>(n, 0);
        for (int i = 0; i < n; i++) fa[i] = i;
    }

    int find(int x) {
        if (x == fa[x]) return x;
        return fa[x] = find(fa[x]);
    }

    void unionSet(int x, int y) {
        x = find(x), y = find(y);
        if (x != y) fa[x] = y;
    }

private:
    vector<int> fa;
};

class Solution {
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        DisjointSet s(isConnected.size());
        for (int i = 0; i < isConnected.size(); i++)
            for (int j = 0; j < isConnected.size(); j++)
                if (isConnected[i][j] == 1) s.unionSet(i, j);
        int ans = 0;
        for (int i = 0; i < isConnected.size(); i++)
            if (s.find(i) == i) ans++;
        return ans;
    }
};

File: 第14课 课件+四种语言模板+代码/【并查集模板】lc547 省份数量.js
class DisjointSet {
    constructor(n) {
        this.fa = [];
        for (let i = 0; i < n; i++) this.fa[i] = i;
    }

    find(x) {
        if (x == this.fa[x]) return x;
        return this.fa[x] = this.find(this.fa[x]);
    }

    unionSet(x, y) {
        x = this.find(x);
        y = this.find(y);
        if (x != y) this.fa[x] = y;
    }
};
/**
 * @param {number[][]} isConnected
 * @return {number}
 */
var findCircleNum = function(isConnected) {
        let s = new DisjointSet(isConnected.length);
        for (let i = 0; i < isConnected.length; i++)
            for (let j = 0; j < isConnected.length; j++)
                if (isConnected[i][j] == 1) s.unionSet(i, j);
        let ans = 0;
        for (let i = 0; i < isConnected.length; i++)
            if (s.find(i) == i) ans++;
        return ans;
};